File: 436. Find Right Interval/C++/Solution.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<int> findRightInterval(vector<Interval> &intervals) {
        vector<int> result(intervals.size());
        vector<int> starts(intervals.size());
        unordered_map<int, int> indexs;
        for (int i = 0; i < intervals.size(); i++) {
            starts[i] = intervals[i].start;
            indexs[intervals[i].start] = i;
        }
        sort(starts.begin(), starts.end());
        for (int i = 0; i < intervals.size(); i++) {
            int end = intervals[i].end;
            int low = 0;
            int high = starts.size() - 1;
            while (low <= high) {
                int mid = (low + high) / 2;
                if (starts[mid] < end) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
            if (low == starts.size()) {
                result[i] = -1;
            } else {
                result[i] = indexs[starts[low]];
            }
        }
        return result;

    }
};

File: 436. Find Right Interval/Python/Solution.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e


class Solution(object):
    def findRightInterval(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[int]
        """
        result = []
        starts = [interval.start for interval in intervals]
        starts.sort()
        indexs = {}
        for index, interval in enumerate(intervals):
            indexs[interval.start] = index
        for interval in intervals:
            end = interval.end
            low = 0
            high = len(starts) - 1
            while low <= high:
                mid = (low + high) // 2
                if starts[mid] < end:
                    low = mid + 1
                else:
                    high = mid - 1
            if low == len(starts):
                result.append(-1)
            else:
                result.append(indexs[starts[low]])
        return result
